_rules/text-node-contrast-afw4f7.md

```diff
@@ -0,0 +1,337 @@
+---
+id: afw4f7
+name: Text node has minimal contrast
+rule_type: atomic
+
+description: |
+  This rule checks that every text node has minimal color contrast with its background
+
+accessibility_requirements:
+  wcag20:1.4.3: # Contrast (Minimum)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+  - Language
+
+authors:
+  - Brian Bors
+  - Kasper Isager
+  - Wilco Fiers
+---
+
+## Applicability
+
+Any [text node](https://dom.spec.whatwg.org/#text) that is [visible](#visible) and a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendent](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:
+
+- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)
+- Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled](#disabled) (i.e. It is part of the label)
+- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled](#disabled)
+
+**Note**: When the text color is the same as the background color, the element is not visible, and so is not applicable.
```

Comment:
I think this note should be moved to the definition of `visible`.

Also, most of our "invisible" examples (especially when it need to be in the accessibility tree) in rules (all rules) tend to be with text out of screen (`style="position: absolute; top: -9999px; left: -9999px;"`). We should probably add some with text color being same as background color.

Reply:
The definition of visible usually has nothing to with differences in colour so it should not be mentioned in that definition.

Reply:
@ShadowBB I do not fully understand. Do you mean that
```html
<p style="color: #000; background: #000;">Some text</p>
```
This is visible and thus should be checked by all rules that have "visible" in the applicability?